src/main/java/com/learning/mfscreener/service/HistoricalNavService.java

```diff
@@ -53,104 +53,135 @@ public HistoricalNavService(
     }
 
     public String getHistoricalNav(Long schemeCode, LocalDate navDate) {
-        // URL https://portal.amfiindia.com/DownloadNAVHistoryReport_Po.aspx?tp=1&frmdt=01-Jan-2024&todt=03-Jan-2024
-        String toDate = navDate.format(FORMATTER_DD_MMM_YYYY);
-        String fromDate = navDate.minusDays(3).format(FORMATTER_DD_MMM_YYYY);
-        // tp=3 Interval Fund Schemes ( Income )
-        // tp=2 Close Ended Schemes ( Income )
-        // tp=1 Open Ended Schemes
-        String historicalUrl = "https://portal.amfiindia.com/DownloadNAVHistoryReport_Po.aspx?frmdt=%s&todt=%s"
-                .formatted(fromDate, toDate);
+        URI historicalNavUri = buildHistoricalNavUri(navDate);
         Optional<MFSchemeEntity> bySchemeId = schemeService.findBySchemeCode(schemeCode);
         String payOut;
         boolean persistSchemeInfo = false;
-        SchemeNameAndISIN firstByAmfi = null;
+        SchemeNameAndISIN schemeNameAndISIN = null;
         if (bySchemeId.isEmpty()) {
             // discontinued Scheme
-            firstByAmfi = userSchemeDetailsService
-                    .findFirstBySchemeCode(schemeCode)
-                    .orElseThrow(
-                            () -> new SchemeNotFoundException("Fund with schemeCode " + schemeCode + " Not Found"));
-            payOut = firstByAmfi.getIsin();
+            schemeNameAndISIN = fetchSchemeDetails(schemeCode);
+            payOut = schemeNameAndISIN.getIsin();
             persistSchemeInfo = true;
         } else {
             payOut = bySchemeId.get().getPayOut();
         }
+        return fetchAndProcessNavData(
+                historicalNavUri, payOut, persistSchemeInfo, schemeNameAndISIN, schemeCode, navDate);
+    }
 
-        URI uri = UriComponentsBuilder.fromHttpUrl(historicalUrl).build().toUri();
-        String oldSchemeId = null;
+    String fetchAndProcessNavData(
+            URI historicalNavUri,
+            String payOut,
+            boolean persistSchemeInfo,
+            SchemeNameAndISIN schemeNameAndISIN,
+            Long schemeCode,
+            LocalDate navDate) {
         try {
-            String allNAVs = restClient.get().uri(uri).retrieve().body(String.class);
-            Reader inputString = new StringReader(Objects.requireNonNull(allNAVs));
-            try (BufferedReader br = new BufferedReader(inputString)) {
-                String lineValue = br.readLine();
-                for (int i = 0; i < 2; ++i) {
+            Reader inputString = fetchHistoricalNavByCallingUri(historicalNavUri);
+            return parseNavData(inputString, payOut, persistSchemeInfo, schemeNameAndISIN, schemeCode, navDate);
+        } catch (ResourceAccessException exception) {
+            // eating as we can't do much, and it should be set when available
+            LOGGER.error("Unable to load Historical Data, downstream service is down ", exception);
+            return null;
+        }
+    }
+
+    String parseNavData(
```

Comment:
@coderabbitai, can you please reduce the cognitive complexity of this method without any change to code logic